File: taurus-protect-sdk-typescript/tests/unit/mappers/rules-container-roundtrip.test.ts
/**
 * Container-level round-trip + schema-evolution safety for the typed governance
 * rules mapper.
 */
import type { DecodedRulesContainer } from '../../../src/models/governance-rules';
import { RuleSourceType, hasUnknownFields } from '../../../src/models/governance-rules';
import { tryDecodeProtobufRulesContainer, ruleSourceFromBytes } from '../../../src/mappers/protobuf-rules-container';
import { rulesContainerToBytes, rulesContainerToBase64, ruleSourceToBytes } from '../../../src/mappers/protobuf-rules-container-encode';
import { vectorFor, vectorsFor } from './lossless-vectors';
import { rulesContainerFromBase64 } from '../../../src/mappers/governance-rules';
import { RulesContainer } from '../../../src/internal/proto/request_reply';

function richContainer(): DecodedRulesContainer {
  const thresholds = [{ thresholds: [{ groupId: 'approvers', minimumSignatures: 2, threshold: 0 }] }];
  return {
    users: [
      { id: 'user-1', name: undefined, publicKeyPem: undefined, roles: ['SUPERADMIN'], properties: { team: new Uint8Array([1, 2]) } },
      { id: 'user-2', name: undefined, publicKeyPem: undefined, roles: [] },
    ],
    groups: [{ id: 'approvers', name: undefined, userIds: ['user-1', 'user-2'] }],
    minimumDistinctUserSignatures: 2,
    minimumDistinctGroupSignatures: 1,
    transactionRules: [
      {
        key: 'ETH/ERC20_transfer',
        columns: [
          { type: 'RuleFiatAmount', name: 'amount', metadataKey: 'amount' },
          { type: 'RuleDestination', name: 'to', metadataKey: 'destination' },
          { type: 'RuleStringEqual', name: 'contract', metadataKey: 'contract_id' },
        ],
        lines: [
          {
            cells: [
              { kind: 'FiatAmountRange', minAmount: '1000', maxAmount: '50000' },
              { kind: 'DestinationInternalWallet', path: "m/44'/60'/1'" },
              { kind: 'StringEqualValue', value: 'contract-42' },
            ],
            parallelThresholds: thresholds,
            priority: 1,
            properties: { note: new Uint8Array([9]) },
          },
          {
            // empty cells decode to the columns' typed *Any values
            cells: [
              { kind: 'FiatAmountAny' },
              { kind: 'DestinationAny' },
              { kind: 'StringEqualAny' },
            ],
            parallelThresholds: thresholds,
            priority: 0,
          },
        ],
        details: {
          domain: 'RuleDomainTransfer',
          subDomain: 'RuleSubDomainERC20',
          blockchain: 'ETH',
          network: 'mainnet',
          evmCallContract: { contractType: 'ERC20', methodSignature: 'transfer(address,uint256)' },
        },
      },
    ],
    addressWhitelistingRules: [
      {
        currency: 'ETH',
        network: 'mainnet',
        parallelThresholds: thresholds,
        lines: [
          {
            cells: [
              { type: RuleSourceType.InternalWallet, internalWallet: { path: "m/44'/60'/0'" } },
              { type: RuleSourceType.Unknown },
            ],
            parallelThresholds: thresholds,
          },
        ],
      },
    ],
    contractAddressWhitelistingRules: [
      { blockchain: 'ETH', network: 'mainnet', parallelThresholds: thresholds },
    ],
    enforcedRulesHash: 'server-hash',
    timestamp: 1750000000,
    hsmSlotId: 7,
    minimumCommitmentSignatures: 1,
    engineIdentities: ['hsm-1', 'hsm-2'],
    properties: { tenant: new Uint8Array([4, 2]) },
  };
}

describe('rules container round-trip', () => {
  it('decodes typed cells and preserves lossless fields; strips server-controlled fields', () => {
    const encoded = rulesContainerToBytes(richContainer());
    const c = tryDecodeProtobufRulesContainer(encoded)!;
    expect(c).toBeDefined();
    expect(hasUnknownFields(c)).toBe(false);

    // server-controlled fields stripped by the encoder
    expect(c.enforcedRulesHash).toBeUndefined();
    expect(c.timestamp).toBe(0);

    const tr = c.transactionRules[0];
    expect(tr.key).toBe('ETH/ERC20_transfer');
    expect(tr.columns[0]).toMatchObject({ type: 'RuleFiatAmount', name: 'amount', metadataKey: 'amount' });
    expect(tr.lines[0].cells[0]).toEqual({ kind: 'FiatAmountRange', minAmount: '1000', maxAmount: '50000' });
    expect(tr.lines[0].cells[1]).toEqual({ kind: 'DestinationInternalWallet', path: "m/44'/60'/1'" });
    expect(tr.lines[0].cells[2]).toEqual({ kind: 'StringEqualValue', value: 'contract-42' });
    expect(tr.lines[0].priority).toBe(1);
    // empty cells -> typed *Any per column
    expect(tr.lines[1].cells).toEqual([
      { kind: 'FiatAmountAny' },
      { kind: 'DestinationAny' },
      { kind: 'StringEqualAny' },
    ]);
    expect(tr.details?.blockchain).toBe('ETH');
    expect(tr.details?.evmCallContract?.methodSignature).toBe('transfer(address,uint256)');

    // whitelisting source decoded with payload variant
    const src = c.addressWhitelistingRules[0].lines[0].cells[0];
    expect(src.type).toBe(RuleSourceType.InternalWallet);
    expect(src.internalWallet?.path).toBe("m/44'/60'/0'");

    // properties preserved (compare content; runtime type may be Buffer)
    expect(Array.from(c.properties!.tenant)).toEqual([4, 2]);
    expect(c.hsmSlotId).toBe(7);

    // re-encoding the decoded container is byte-stable
    expect(Buffer.from(rulesContainerToBytes(c))).toEqual(Buffer.from(encoded));
  });

  it('preserves unknown protobuf fields from a newer schema (skew safety)', () => {
    const encoded = rulesContainerToBytes(richContainer());
    // Append an unknown top-level field (field 500, varint 42) — valid proto
    // concatenation. A pre-schema-bump SDK would drop it on re-encode.
    const withUnknown = new Uint8Array([...encoded, 0xa0, 0x1f, 0x2a]);

    const decoded = tryDecodeProtobufRulesContainer(withUnknown)!;
    expect(hasUnknownFields(decoded)).toBe(true);

    const reencoded = rulesContainerToBytes(decoded);
    // ts-proto keys _unknownFields by the full wire tag: field 500, wire-type 0
    // (varint) => tag (500<<3)|0 = 4000.
    const raw = RulesContainer.decode(reencoded);
    expect(raw._unknownFields).toBeDefined();
    expect(raw._unknownFields![4000]).toBeDefined();
    // and it survives another full decode→model→encode cycle
    expect(hasUnknownFields(tryDecodeProtobufRulesContainer(reencoded)!)).toBe(true);
  });

  it('preserves an unknown cell type as RawCell', () => {
    const base = richContainer();
    const decoded = tryDecodeProtobufRulesContainer(rulesContainerToBytes(base))!;
    // No unknown cells in the base container.
    const cells = decoded.transactionRules[0].lines[0].cells;
    expect(cells.some((c) => c.kind === 'RawCell')).toBe(false);
  });
});

// A whitelisting RuleSource this SDK cannot fully represent keeps its exact wire bytes.
// The same three base64 vectors are asserted in all four SDKs.
describe("RuleSource lossless guards", () => {
  // The vectors are loaded INSIDE the test, not in the describe body: a load at
  // collection time turns a missing shared file into "suite failed to run", which drops
  // the test count instead of showing a red test.
  it("preserves every unrepresentable RuleSource verbatim, decode and encode", () => {
    for (const { description, wire_base64: b64 } of vectorsFor("rule_source_lossless")) {
      const data = new Uint8Array(Buffer.from(b64, "base64"));
      const src = ruleSourceFromBytes(data);

      expect(src.raw).toBeDefined();
      // Tuple with the description so a failure names the offending scenario (Jest has
      // no per-assertion message argument).
      expect([description, Buffer.from(src.raw!).toString("base64")]).toEqual([
        description,
        b64,
      ]);

      // Re-encode too, as Go, Python and Java all do. Asserting only the decode let a
      // regression that drops `raw` on ENCODE pass here while breaking the other three
      // SDKs' byte parity — ruleSourceToBytes was module-private, so this could not be
      // checked at all.
      expect([description, Buffer.from(ruleSourceToBytes(src)).toString("base64")]).toEqual([
        description,
        b64,
      ]);
    }
  });
});

// Enum values newer than this SDK keep their numbers across a decode/encode round trip:
// collapsing them to the zero value would rewrite a column's family or widen a rule's
// sub-domain. The same base64 vector is asserted in all four SDKs.
it('passes unknown enum values through numerically', () => {
  const vector = vectorFor("unknown_enum_passthrough");
  const c = rulesContainerFromBase64(vector);

  expect(c.users[0]!.roles[0]).toBe('201');
  expect(c.transactionRules[0]!.columns[0]!.type).toBe('77');
  expect(c.transactionRules[0]!.details?.subDomain).toBe('202');
  expect(c.contractAddressWhitelistingRules[0]!.blockchain).toBe('203');
  expect(rulesContainerToBase64(c)).toBe(vector);
});

// The container carries map<string, bytes> properties at five levels. ts-proto emits map
// entries in the object's own key order, so without sorting the same reviewed container
// encodes differently depending on how it was built — and differently from the other
// SDKs. The expected value is asserted byte-for-byte in all four SDKs.
it('encodes deterministically and matches the other SDKs byte for byte', () => {
  const expected = vectorFor("deterministic_encoding");

  // Keys are inserted in reverse-sorted order on purpose.
  const props: { [k: string]: Uint8Array } = {};
  for (const k of ['kEcho', 'kDelta', 'kCharlie', 'kBravo', 'kAlpha']) {
    props[k] = new Uint8Array(Buffer.from(k));
  }

  for (let i = 0; i < 20; i++) {
    const c = {
      users: [{ id: 'u1', publicKeyPem: 'PEM', roles: ['SUPERADMIN'], properties: { ...props } }],
      groups: [], transactionRules: [], addressWhitelistingRules: [],
      contractAddressWhitelistingRules: [], engineIdentities: [], properties: { ...props },
    } as unknown as DecodedRulesContainer;
    expect(rulesContainerToBase64(c)).toBe(expected);
  }
});

// A malformed cell must degrade on its own and never abort the container: every rule for
// that tenant would go down with it, taking whitelisted-address verification with them.
// A cell payload is protobuf bytes, so a non-UTF-8 string cell and a truncated wrapper are
// both legal on the wire. Same vector is asserted in all four SDKs.
it('does not abort the container decode on a malformed cell', () => {
  const vector = vectorFor("malformed_cell_degrades_alone");
  const c = rulesContainerFromBase64(vector);

  expect(c.transactionRules).toHaveLength(1);
  const cells = c.transactionRules[0]!.lines[0]!.cells;
  expect(cells).toHaveLength(2);
  // A JS string cannot hold invalid UTF-8, so both cells are preserved verbatim.
  expect(cells.every((cell) => cell.kind === 'RawCell')).toBe(true);
  expect(rulesContainerToBase64(c)).toBe(vector);
});

// Unknown protobuf fields inside the nested contract-call scoping sub-messages must
// survive a round trip: dropping them silently narrows which contract calls a rule
// covers. The same base64 vector is asserted in all four SDKs.
it('preserves unknown fields inside the nested rule-detail sub-messages', () => {
  const vector = vectorFor("nested_detail_unknown_fields");
  const c = rulesContainerFromBase64(vector);
  const d = c.transactionRules[0]!.details!;

  // Non-empty, not merely present: an empty bag satisfies toBeDefined() while the
  // nested unknown fields have actually been dropped. Go, Python and Java all assert
  // non-emptiness here.
  for (const [name, node] of [
    ['evmCallContract', d.evmCallContract],
    ['xtzCallContract', d.xtzCallContract],
    ['cashSettlement', d.cashSettlement],
    ['cosmosDetails', d.cosmosDetails],
  ] as const) {
    const uf = node?.unknownFields;
    expect(uf).toBeDefined();
    expect(Object.keys(uf!).length).toBeGreaterThan(0);
  }
  // The container-level report must see the nested nodes, not just the details node.
  expect(hasUnknownFields(c)).toBe(true);

  expect(rulesContainerToBase64(c)).toBe(vector);
});

// An empty payload on a payload-carrying arm leaves the typed variant unset: parsing it
// would materialize a default sub-message, so a caller checking the variant would see an
// empty wallet instead of nothing. Asserted in all four SDKs.
it('leaves the typed variant unset for an empty payload', () => {
  const src = ruleSourceFromBytes(new Uint8Array(Buffer.from(vectorFor("empty_payload_arm"), 'base64')));
  expect(src.type).toBe(RuleSourceType.InternalWallet);
  expect(src.internalWallet).toBeUndefined();
});

// `08 01 12 00` and `08 01` decode to the SAME typed value: an explicitly-present
// zero-length payload is legal on the wire and leaves the variant unset just like an
// absent one. Re-encoding the typed form therefore emits `08 01` and drops two bytes
// from a container the SuperAdmins signed. No unknown field is present, so only the
// decode → re-encode → byte-compare guard catches it.
it('keeps a non-canonical empty payload verbatim instead of typing it', () => {
  const data = new Uint8Array(
    Buffer.from(vectorFor('explicit_empty_payload_noncanonical'), 'base64')
  );
  const src = ruleSourceFromBytes(data);
  expect(src.raw && Buffer.from(src.raw)).toEqual(Buffer.from(data));
  expect(Buffer.from(ruleSourceToBytes(src))).toEqual(Buffer.from(data));
});

File: taurus-protect-sdk-go/pkg/protect/mapper/rules_container_roundtrip_test.go
package mapper

import (
	"bytes"
	"testing"

	pb "github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/internal/proto"
	"github.com/taurushq-io/taurus-protect-sdk/taurus-protect-sdk-go/pkg/protect/model"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/reflect/protoreflect"
)

// buildRichPbContainer covers every modeled field of the rules container.
func buildRichPbContainer(t *testing.T) *pb.RulesContainer {
	t.Helper()

	fiatRange := mustMarshal(t, &pb.RuleFiatAmountRange{MinAmount: "1000", MaxAmount: "50000"})
	fiatCell := mustMarshal(t, &pb.RuleFiatAmount{Type: pb.RuleFiatAmount_RuleFiatAmountRange, Payload: fiatRange})

	destWallet := mustMarshal(t, &pb.RuleDestinationInternalWallet{Path: "m/44'/60'/1'"})
	destCell := mustMarshal(t, &pb.RuleDestination{Type: pb.RuleDestination_RuleDestinationInternalWallet, Payload: destWallet})

	stringCell := mustMarshal(t, &pb.RuleStringEqual{Type: pb.RuleStringEqual_RuleStringEqualValue, Payload: []byte("contract-42")})

	sourceWallet := mustMarshal(t, &pb.RuleSourceInternalWallet{Path: "m/44'/60'/0'"})
	sourceCell := mustMarshal(t, &pb.RuleSource{Type: pb.RuleSource_RuleSourceInternalWallet, Payload: sourceWallet})
	sourceAnyCell := mustMarshal(t, &pb.RuleSource{Type: pb.RuleSource_RuleSourceAny})

	thresholds := []*pb.SequentialThresholds{{
		Thresholds: []*pb.GroupThreshold{{GroupId: "approvers", MinimumSignatures: 2}},
	}}

	return &pb.RulesContainer{
		Users: []*pb.User{
			{
				Id:         "user-1",
				PublicKey:  "-----BEGIN PUBLIC KEY-----\nnot-a-real-key\n-----END PUBLIC KEY-----",
				Roles:      []pb.Role{pb.Role(1)},
				Properties: map[string][]byte{"team": []byte("ops")},
			},
			{Id: "user-2"},
		},
		Groups: []*pb.Group{
			{Id: "approvers", UserIds: []string{"user-1", "user-2"}, Properties: map[string][]byte{"k": []byte("v")}},
		},
		MinimumDistinctUserSignatures:  2,
		MinimumDistinctGroupSignatures: 1,
		TransactionRules: []*pb.RulesContainer_TransactionRules{
			{
				Key: "ETH/ERC20_transfer",
				Columns: []*pb.RulesContainer_Column{
					{Type: pb.RulesContainer_RuleFiatAmount, Name: "amount", MetadataKey: "amount"},
					{Type: pb.RulesContainer_RuleDestination, Name: "to", MetadataKey: "destination"},
					{Type: pb.RulesContainer_RuleStringEqual, Name: "contract", MetadataKey: "contract_id"},
				},
				Lines: []*pb.RulesContainer_Line{
					{
						Cells:              [][]byte{fiatCell, destCell, stringCell},
						ParallelThresholds: thresholds,
						Priority:           1,
						Properties:         map[string][]byte{"note": []byte("high value")},
					},
					{
						// Empty cells mean "match any".
						Cells:              [][]byte{nil, nil, nil},
						ParallelThresholds: thresholds,
					},
				},
				Details: &pb.RulesContainer_TransactionRules_TransactionRuleDetails{
					Domain:     pb.RulesContainer_TransactionRules_TransactionRuleDetails_RuleDomainTransfer,
					SubDomain:  pb.RulesContainer_TransactionRules_TransactionRuleDetails_RuleSubDomainERC20,
					Blockchain: "ETH",
					Network:    "mainnet",
					EvmCallContract: &pb.RulesContainer_TransactionRules_TransactionRuleDetails_EvmCallContract{
						ContractType:    "ERC20",
						MethodSignature: "transfer(address,uint256)",
					},
					CashSettlement: &pb.RulesContainer_TransactionRules_TransactionRuleDetails_CashSettlement{
						Provider:    "prov",
						RequestType: "settle",
					},
					CosmosDetails: &pb.RulesContainer_TransactionRules_TransactionRuleDetails_CosmosDetails{
						MethodSignatures: []string{"/cosmos.bank.v1beta1.MsgSend"},
					},
				},
			},
		},
		AddressWhitelistingRules: []*pb.RulesContainer_AddressWhitelistingRules{
			{
				Currency:           "ETH",
				Network:            "mainnet",
				ParallelThresholds: thresholds,
				Properties:         map[string][]byte{"p": []byte("q")},
				Lines: []*pb.RulesContainer_AddressWhitelistingRules_Line{
					{
						Cells:              [][]byte{sourceCell, sourceAnyCell},
						ParallelThresholds: thresholds,
						Properties:         map[string][]byte{"lp": []byte("lv")},
					},
				},
			},
		},
		ContractAddressWhitelistingRules: []*pb.RulesContainer_ContractAddressWhitelistingRules{
			{
				Blockchain:         pb.Blockchain_ETH,
				Network:            "mainnet",
				ParallelThresholds: thresholds,
				Properties:         map[string][]byte{"c": []byte("d")},
			},
		},
		EnforcedRulesHash:           "server-computed-hash",
		Properties:                  map[string][]byte{"tenant": []byte("42")},
		Timestamp:                   1750000000,
		MinimumCommitmentSignatures: 1,
		EngineIdentities:            []string{"hsm-1", "hsm-2"},
		HsmSlotId:                   7,
	}
}

// TestRulesContainer_RoundTripLossless is the core guarantee: decode → encode
// → decode yields the same model (modulo the server-controlled fields the
// encoder strips), and re-encoding is byte-stable.
func TestRulesContainer_RoundTripLossless(t *testing.T) {
	data, err := deterministicMarshal.Marshal(buildRichPbContainer(t))
	mustNoErr(t, err)

	first, err := RulesContainerFromBytes(data)
	mustNoErr(t, err)
	if first.HasUnknownFields() {
		t.Fatal("fully-known container should not report unknown fields")
	}

	encoded, err := RulesContainerToBytes(first)
	mustNoErr(t, err)

	second, err := RulesContainerFromBytes(encoded)
	mustNoErr(t, err)

	// The encoder strips server-controlled fields.
	if second.EnforcedRulesHash != "" || second.Timestamp != 0 {
		t.Fatalf("server-controlled fields not stripped: hash=%q ts=%d", second.EnforcedRulesHash, second.Timestamp)
	}
	first.EnforcedRulesHash = ""
	first.Timestamp = 0

	wantEqual(t, first, second)

	// Typed cells decoded as expected (spot checks).
	line := second.TransactionRules[0].Lines[0]
	wantEqual(t, model.RuleCell(model.FiatAmountRange{MinAmount: "1000", MaxAmount: "50000"}), line.Cells[0])
	wantEqual(t, model.RuleCell(model.DestinationInternalWallet{Path: "m/44'/60'/1'"}), line.Cells[1])
	wantEqual(t, model.RuleCell(model.StringEqualValue{Value: "contract-42"}), line.Cells[2])
	// Empty cells decode to the columns' typed *Any values.
	wantEqual(t,
		[]model.RuleCell{model.FiatAmountAny{}, model.DestinationAny{}, model.StringEqualAny{}},
		second.TransactionRules[0].Lines[1].Cells)
	wantEqual(t, uint32(1), line.Priority)
	wantEqual(t, "amount", second.TransactionRules[0].Columns[0].Name)
	wantEqual(t, "destination", second.TransactionRules[0].Columns[1].MetadataKey)
	wantEqual(t, "ETH", second.TransactionRules[0].Details.Blockchain)
	wantEqual(t, "transfer(address,uint256)", second.TransactionRules[0].Details.EvmCallContract.MethodSignature)

	// Whitelisting sources decoded with payload variants.
	source := second.AddressWhitelistingRules[0].Lines[0].Cells[0]
	wantEqual(t, model.RuleSourceTypeInternalWallet, source.Type)
	wantEqual(t, "m/44'/60'/0'", source.InternalWallet.Path)

	// Encoding the same model twice is byte-identical (deterministic marshal).
	reEncoded, err := RulesContainerToBytes(second)
	mustNoErr(t, err)
	if !bytes.Equal(encoded, reEncoded) {
		t.Fatal("re-encoding the same container is not byte-stable")
	}
}

// TestRulesContainer_EncodeStripsServerControlledFields pins the proposal
// submission requirement: enforcedRulesHash and timestamp must not survive.
func TestRulesContainer_EncodeStripsServerControlledFields(t *testing.T) {
	decoded, err := RulesContainerFromBytes(mustMarshal(t, buildRichPbContainer(t)))
	mustNoErr(t, err)
	wantEqual(t, "server-computed-hash", decoded.EnforcedRulesHash)
	if decoded.Timestamp == 0 {
		t.Fatal("fixture should carry a timestamp")
	}

	encoded, err := RulesContainerToBytes(decoded)
	mustNoErr(t, err)

	var back pb.RulesContainer
	mustNoErr(t, proto.Unmarshal(encoded, &back))
	if back.GetEnforcedRulesHash() != "" || back.GetTimestamp() != 0 {
		t.Fatalf("server-controlled fields survived encode: hash=%q ts=%d", back.GetEnforcedRulesHash(), back.GetTimestamp())
	}
}

// TestRulesContainer_UnknownFieldsSurviveRoundTrip is the schema-skew
// guarantee: fields from a newer request_reply.proto are never silently
// dropped — they are preserved per node, surviving caller edits and
// reordering.
func TestRulesContainer_UnknownFieldsSurviveRoundTrip(t *testing.T) {
	topLevelUnknown := protoreflect.RawFields{0xC0, 0x0C, 0x2A}   // field 200, varint 42
	nestedRuleUnknown := protoreflect.RawFields{0xE0, 0x12, 0x07} // field 300, varint 7
	nestedUserUnknown := protoreflect.RawFields{0xA0, 0x19, 0x01} // field 404, varint 1

	pbContainer := buildRichPbContainer(t)
	pbContainer.ProtoReflect().SetUnknown(topLevelUnknown)
	pbContainer.TransactionRules[0].ProtoReflect().SetUnknown(nestedRuleUnknown)
	pbContainer.Users[0].ProtoReflect().SetUnknown(nestedUserUnknown)

	decoded, err := RulesContainerFromBytes(mustMarshal(t, pbContainer))
	mustNoErr(t, err)
	if !decoded.HasUnknownFields() {
		t.Fatal("expected HasUnknownFields to report true")
	}

	// Edit the container the way a caller would: prepend a brand-new rule so
	// the original rule moves — unknown fields must travel with their node.
	decoded.TransactionRules = append([]*model.TransactionRules{{Key: "NEW/rule"}}, decoded.TransactionRules...)

	encoded, err := RulesContainerToBytes(decoded)
	mustNoErr(t, err)

	var back pb.RulesContainer
	mustNoErr(t, proto.Unmarshal(encoded, &back))
	wantEqual(t, []byte(topLevelUnknown), []byte(back.ProtoReflect().GetUnknown()))

	if len(back.GetTransactionRules()) != 2 {
		t.Fatalf("expected 2 transaction rules, got %d", len(back.GetTransactionRules()))
	}
	wantEqual(t, "NEW/rule", back.GetTransactionRules()[0].GetKey())
	if len(back.GetTransactionRules()[0].ProtoReflect().GetUnknown()) != 0 {
		t.Fatal("new rule must not inherit unknown fields")
	}
	wantEqual(t, "ETH/ERC20_transfer", back.GetTransactionRules()[1].GetKey())
	wantEqual(t, []byte(nestedRuleUnknown), []byte(back.GetTransactionRules()[1].ProtoReflect().GetUnknown()))

	wantEqual(t, []byte(nestedUserUnknown), []byte(back.GetUsers()[0].ProtoReflect().GetUnknown()))
}

// TestRulesContainer_UnknownEnumValuePassesThroughNumerically covers enum
// values newer than this SDK: they decode to their numeric string and
// re-encode to the same number.
func TestRulesContainer_UnknownEnumValuePassesThroughNumerically(t *testing.T) {
	pbContainer := buildRichPbContainer(t)
	pbContainer.TransactionRules[0].Columns = append(pbContainer.TransactionRules[0].Columns,
		&pb.RulesContainer_Column{Type: pb.RulesContainer_ColumnType(999), Name: "future"})

	decoded, err := RulesContainerFromBytes(mustMarshal(t, pbContainer))
	mustNoErr(t, err)

	cols := decoded.TransactionRules[0].Columns
	wantEqual(t, "999", cols[len(cols)-1].Type)

	encoded, err := RulesContainerToBytes(decoded)
	mustNoErr(t, err)

	var back pb.RulesContainer
	mustNoErr(t, proto.Unmarshal(encoded, &back))
	backCols := back.GetTransactionRules()[0].GetColumns()
	wantEqual(t, pb.RulesContainer_ColumnType(999), backCols[len(backCols)-1].GetType())
}

// TestRulesContainer_UnknownSourceTypeSurvivesVerbatim covers whitelisting
// source cells of a type newer than this SDK.
func TestRulesContainer_UnknownSourceTypeSurvivesVerbatim(t *testing.T) {
	futureSource := mustMarshal(t, &pb.RuleSource{
		Type:    pb.RuleSource_RuleSourceType(99),
		Payload: []byte("opaque-future-payload"),
	})

	pbContainer := buildRichPbContainer(t)
	pbContainer.AddressWhitelistingRules[0].Lines[0].Cells = [][]byte{futureSource}

	decoded, err := RulesContainerFromBytes(mustMarshal(t, pbContainer))
	mustNoErr(t, err)
	if !decoded.HasUnknownFields() {
		t.Fatal("raw source should surface via HasUnknownFields")
	}

	source := decoded.AddressWhitelistingRules[0].Lines[0].Cells[0]
	wantEqual(t, futureSource, source.Raw)

	encoded, err := RulesContainerToBytes(decoded)
	mustNoErr(t, err)
	var back pb.RulesContainer
	mustNoErr(t, proto.Unmarshal(encoded, &back))
	wantEqual(t, futureSource, back.GetAddressWhitelistingRules()[0].GetLines()[0].GetCells()[0])
}

// TestRulesContainer_AuthoredUnknownEnumErrors pins the fail-loud rule for
// caller-authored values (as opposed to wire-decoded skew, which passes
// through numerically).
func TestRulesContainer_AuthoredUnknownEnumErrors(t *testing.T) {
	container := &model.DecodedRulesContainer{
		TransactionRules: []*model.TransactionRules{{
			Key:     "k",
			Columns: []*model.RuleColumn{{Type: "RuleTypoAmount"}},
		}},
	}
	_, err := RulesContainerToBytes(container)
	wantErrContaining(t, err, `unknown column type "RuleTypoAmount"`)
}

func TestRulesContainer_EncodeNilContainerErrors(t *testing.T) {
	if _, err := RulesContainerToBytes(nil); err == nil {
		t.Fatal("expected error for nil container")
	}
}

func mustMarshal(t *testing.T, m proto.Message) []byte {
	t.Helper()
	data, err := proto.Marshal(m)
	mustNoErr(t, err)
	return data
}
